### PR TITLE
Resetpassword2

### PR DESCRIPTION
**Pull request by noodlemoodle**
moved from https://github.com/kgrgreer/foam3/pull/2104
now support duplicate email checking

nanopay update https://github.com/nanoPayinc/NANOPAY/pull/15462

### CODE REVIEW

File: src/foam/nanos/auth/email/ServerEmailVerificationService.js [hunk 1]
```diff
@@ -31,12 +36,41 @@
   ],
 
   methods: [
+    {
+      name: 'findUser',
+      args: 'Context x, String email, String userName',
+      type: 'foam.nanos.auth.User',
+      javaCode: `
+        DAO userDAO = ((DAO) x.get("localUserDAO")).where(
```
Comment (kgrgreer) [inline src/foam/nanos/auth/email/ServerEmailVerificationService.js:44]: You could add a limit(2) for a slight performance improvement sometimes.

File: src/foam/nanos/auth/resetPassword/ResetPasswordByCode.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    { name: 'CODE_SECTION_TITLE', message: 'Please enter your email verification code' },
+    { name: 'RESET_PASSWORD_TITLE', message: 'Reset your password' },
+    { name: 'RESET_PASSWORD_SUBTITLE', message: 'Create a new password for your account' }
+  ],
+
+  sections: [
+    {
+      name: 'verificationCodeSection',
+      title: function() {
+        return this.CODE_SECTION_TITLE;
```
Comment (kgrgreer) [inline src/foam/nanos/auth/resetPassword/ResetPasswordByCode.js]: Is this needed? Can't we set section titles directly from the locale system, just like property labels?

File: src/foam/nanos/auth/resetPassword/ServerResetPasswordService.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    'static foam.mlang.MLang.*'
+  ],
+
+  methods: [
+    {
+      name: 'findUser',
+      args: 'Context x, String email, String userName',
+      type: 'foam.nanos.auth.User',
+      javaCode: `
+        DAO userDAO = ((DAO) x.get("localUserDAO")).where(
```
Comment (kgrgreer) [inline src/foam/nanos/auth/resetPassword/ServerResetPasswordService.js:33]: Could add limit(2).

**Comment by jlhughes** [comment]
Where in this does username come from?  Does the password reset form require username if that theme/spid requires it?

**Comment by noodlemoodle** [comment]
if only email is entered and we find it to be a duplicate email, a duplicateemailexception is thrown and the form then asks for the username, same as how our current resetbytoken flow/signin flow handles duplicate emails @jlhughes

**Review by jlhughes — APPROVED** [review]

**Pull request merged**